rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,149 @@
+
+# Basic example
+
+When Kibana loads initially, there is a Newsfeed component in the UI that
+checks internally cached records to see if it must fetch the Elastic News
+Service for newer items. When the Screenshot Mode Service is implemented, the
+Newsfeed component has a source of information to check on whether or not it
+should load in the Kibana UI. If it can avoid loading, it avoids an unnecessary
+HTTP round trip, which weigh heavily on performance.
+
+# Motivation
+
+Kibana Reporting is a commercial feature in Elastic and is highly capable of
+loading the application pages and converting them into a screenshot for PNG or
+PDF export. However, the way it works has downsides with performance,
+maintainability, and expanding it as a tool that can power higher-level
+features. The solution to those downsides is to have Kibana pages themselves
+become more capable at presenting themselves for screenshot capture reports.
+With the Screenshot Mode Service available, Reporting could drop the
+task that it currently has which hurt performance: wasted rendering that is
+replaced with custom styles that make the page "reportable."
+
+The technical advantage of having such as service also leads to making Kibana
+application pages "printable", in the sense that sending the page to a printer
+for a hard copy results in something more meaningful and specialized for paper
+than today's Kibana can guarantee. This isn't a big concern for Kibana since
+there isn't the expectation to improve printing Kibana, but that technical
+direction is appropriate for improving PDF report generation.
+
+# Detailed design
+
+The Screenshot Mode Service is a callable API available as dependency for
+Kibana applications.
+
+A method of the API tells the Application whether or not it should render
+itself to optimize for non-interactivity.
+
+In a future phase, the API might also tell the application more about the
+report, such as PDF page dimensions, or special layout types (print layout,
+etc).
+
+## Interface
+The `setupDeps.screenshotMode` object has a single purpose: tell the app if it
+should render in an optimized way for screenshot capture:
+
+```
+interface IScreenshotModeServiceSetup {
+  isScreenshotMode: () => boolean;
+}
+```
+
+Internally, this object is constructed from a class that refers to information
+sent via a custom proprietary header:
+
+```
+interface HeaderData {
+  'X-Screenshot-Mode': true
+}
+
+class ScreenshotModeServiceSetup implements IScreenshotModeServiceSetup {
+  constructor(rawData: HeaderData) {}
+  public isScreenshotMode (): boolean {}
+}
+```
+
+This works because the headless browser that opens the page can inject custom
+headers into the request. Teams can test how their app renders when loaded with
+this header using a new configuration setting, or a web debugging proxy, or
+some other tool that is TBD.
+
+# Alternatives
+
+- Print media query CSS
+  If applications UIs supported printability using `@media print`, and Kibana 
```

Comment:
How will we apply the styles ? i think we should use print media queries to apply different print styling, rather than always use js (we could conditionally add custom classes to elements). I think it makes sense using the standard for what's available and only use our custom service to solve the performance problems.

Reply:
The URL of the page has it's own flag (defined by the application) to render a "print-friendly" mode, and the page is responsible for applying those styles on its own. Using print media queries could be a future option, but I don't think it is very optimal:

We could create a new export type in Reporting that calls `page.print()`, however I don't think that is the optimal flow:
1. Load and render the entire page and wait for the network to idle.
2. Call `page.print()`
3. Entire page re-renders as "print" media for final capture.

Step 1 wasted time with rendering the entire page the page with as "screen" media.

In this RFC proposal, the application should send a custom URL to Reporting where "screen" styles look the same as "print" media. It would make sense for the application to re-use those styles in their own print media CSS.

Reply:
The two stated goals of this RFC are to solve the ownership issue and performance issues.

The ownership issue seems like a clear problem to me which is the most urgent to solve. Are the performance issues of equal importance/urgency for us today? (Aside from the fact that we should always be optimizing for performance wherever possible)

The reason I ask is that this proposed alternative feels like a more pragmatic approach IMHO, even if it is missing the performance benefits. Asking apps to create a separate "print" view gives them more flexibility in terms of the implementation, and also makes testing easier as everything is isolated to a dedicated page. Whereas applying a bunch of conditionals around `isScreenshotMode` inside an app's main rendering logic feels like a pretty big ask for teams, which could also be quite risky & error-prone.

> In this RFC proposal, the application should send a custom URL to Reporting where "screen" styles look the same as "print" media.

How would this look? I don't see this interface explicitly mentioned in the RFC, but it feels like a more intuitive approach to me.

Having apps provide their URLs for reporting would also allow them to choose if they want to use `@media print` styles (in which case they just register their normal app URL and maybe they need to tell reporting whether to call `page.print()`), a dedicated page (which could still have the added perf benefits), or something in between (like a url parameter if they want to conditionally change the logic on their existing page). It also places the responsibility for performance on the app teams... which could be viewed as either a good or a bad thing, depending on your perspective 😉

Reply:
Thanks Luke! I think parts of the RFC need to be tweaking to be more clear, that the Screenshot service is primarily for low-level plugins. :sweat_smile: That obviously was only clear in my mind.

> Are the performance issues of equal importance/urgency for us today? 

Yes, and this RFC would solve performance issues of low-level plugins that the app can't turn off, but should be turned off somehow. I mentioned "print media CSS" as another solution that **wouldn't** be able to turn off low-level plugins. 

> Whereas applying a bunch of conditionals around isScreenshotMode inside an app's main rendering logic feels like a pretty big ask for teams, which could also be quite risky & error-prone.

You are right and that is not how we would recommend that the service is used. The app's main rendering logic should be controlled by URL customizations: see Canvas as an example of this working well.

Aside from apps, Kibana also has plugins that run on every page, such as Telemetry. They are not explicitly called by the app, so they need some way to know that they should hold off on rendering when it is screenshot mode.

The service will also help UI dependencies like toast notifications. If the toasts service was aware of screenshot mode, the app can still use it. The toasts service could have options on what to do for screenshots: it could choose to not be visible, or change itself so its visible but doesn't cover up any visualization, etc.

> > In this RFC proposal, the application should send a custom URL to Reporting where "screen" styles look the same as "print" media.

> How would this look? I don't see this interface explicitly mentioned in the RFC, but it feels like a more intuitive approach to me.

That part is the responsibility of the app, and goes along with "The app's main rendering logic should be controlled by URL customizations"

Because the wording wasn't clear, we ended up comparing two concepts:
 - custom URLs controlled by the app for the reporting layout
 - a service to provide an isScreenshotMode flag

They are two different tools in the toolbox. The custom URLs do 80% of the work, and it's a solution already in use by Canvas. The service is a scalpel that aims to solve performance issues with the lower-level stuff.

Reply:
I changed the wording quite a bit to make clear

1. We recommend applications support a customized URL for screenshot capture, and that solves rendering issues for Reporting
2. The Screenshot mode service is a low-level service used by low-level UI plugins that can't be turned off by the application